### ISSUE TITLE
Regarding the "for numerical stability" block

### ISSUE THREAD

**Issue by AlexMoreo** [HobbitLong/SupContrast#65]
I got confused with:
```python
       # compute logits
        anchor_dot_contrast = torch.div(
            torch.matmul(anchor_feature, contrast_feature.T),
            self.temperature)
        # for numerical stability
        logits_max, _ = torch.max(anchor_dot_contrast, dim=1, keepdim=True)
        logits = anchor_dot_contrast - logits_max.detach()
```
Since the max across each row in anchor_dot_contrast will be attained by what you later refer to as "self-contrast cases". These will always be equal to 1/temperature since the features are expected to be l2-normalized. Would it not be better to perform this operation after the mask has been applied?
Thanks for the great work.

**Comment by yash0307** [comment]
@AlexMoreo, did you try this? Does it help?

**Comment by AlexMoreo** [comment]
Hi @yash0307! No, sorry, I don't remember to have ever tried this. Let us know if this happens to work better!

**Comment by adv010** [comment]
@TonyWangX great explanation! Have you considered the reason for normalizing the projection features and the encoder features? From the codebase I'm uncertain whether the projections features do get normalized or not.
